master/buildbot/reporters/hipchat.py

```diff
@@ -0,0 +1,93 @@
+from twisted.python import log
+from twisted.internet import defer
+from buildbot import config
+from buildbot.reporters import utils
+from buildbot.process.results import statusToString
+from buildbot.reporters.http import HttpStatusPushBase
+
+
+class HipChatStatusPush(HttpStatusPushBase):
+    name = "HipChatStatusPush"
+
+    def checkConfig(self, auth_token, endpoint="https://api.hipchat.com",
+                    builder_room_map=None, builder_user_map=None,
+                    event_messages=None, **kwargs):
+        if not isinstance(auth_token, basestring):
+            config.error('auth_token must be a string')
+        if not isinstance(endpoint, basestring):
+            config.error('endpoint must be a string')
+        if builder_room_map and not isinstance(builder_room_map, dict):
+            config.error('builder_room_map must be a dict')
+        if builder_user_map and not isinstance(builder_user_map, dict):
+            config.error('builder_user_map must be a dict')
+
+    @defer.inlineCallbacks
+    def reconfigService(self, auth_token, endpoint="https://api.hipchat.com",
+                        builder_room_map=None, builder_user_map=None,
+                        event_messages=None, **kwargs):
+        yield HttpStatusPushBase.reconfigService(self, **kwargs)
+        self.auth_token = auth_token
+        self.endpoint = endpoint
+        self.builder_room_map = builder_room_map
+        self.builder_user_map = builder_user_map
+        self.user_notify = '%sv2/user/%s/message?auth_token=%s'
+        self.room_notify = '%sv2/room/%s/notification?auth_token=%s'
+
+    @defer.inlineCallbacks
+    def buildStarted(self, key, build):
+        yield self.getBuildDetailsAndSendMessage(build, key[2])
+
+    @defer.inlineCallbacks
+    def buildFinished(self, key, build):
+        yield self.getBuildDetailsAndSendMessage(build, key[2])
+
+    @defer.inlineCallbacks
+    def getBuildDetailsAndSendMessage(self, build, key):
+        yield utils.getDetailsForBuild(self.master, build, **self.neededDetails)
+        postData = yield self.getRecipientList(build, key)
+        postData['message'] = yield self.getMessage(build, key)
+        extra_params = yield self.getExtraParams(build, key)
+        postData.update(extra_params)
+        yield self.send(postData)
+
+    def getRecipientList(self, build, event_name):
+        result = {}
```

Comment:
I think this is good. but we should document it. and also document that it can returned a deferred